Scattered error case, bounds check, and memory leak cleanups

AdamNorberg opened the pull request:
While making libdmg-hfsplus work for packaging Chrome, I noticed an assortment of missing bounds checks, missed `free` calls (especially in error cases), missed error checks, and a recurring pattern of pointer reinterpretation within the HFSPlusCatalogRecord family that was consistently optimistic about assuming it would only get either of two kinds of records. I cleaned up what I ran across when it seemed convenient enough to clean up _or_ bothered me enough that I pursued a larger cleanup.

This PR is neither focused nor comprehensive -- there are issues like this throughout this codebase -- but these are the ones I happened to fix this time. This is built on top of pull requests #17, #18, and #19; the first commit specific to _this_ PR is titled `First bounds check on fullName`.

AdamNorberg committed:
First bounds check on fullName.

AdamNorberg committed:
Wrap "check and cast" pattern for HFSPlusCatalogFolder.

If we use this literally everywhere, it might make it easier to convert to a union-based approach that does not violate the strict aliasing rule. That is probably unreasonably ambitious. However, I would rather have these helpers available instead of repeating the check everywhere, and am likely to rewrite the pattern where I see it, especially if I see repeated casts of the same (ambiguous) pointer, so here it is for my personal convenience.

AdamNorberg committed:
Add bounds check to appending items in addAllInFolder. Fix hfslib.c include syntax.

AdamNorberg committed:
Clean up addAllInFolder overwrite case.

This uses the "check and cast" helpers and adds error checks to handle
unexpected record subtypes. The result is longer but more legible; I
feel like the original was past the line of what is reasonable to
do with the selection operator.

AdamNorberg committed:
Clean up a smattering of memory leaks and missing error checks.

AdamNorberg committed:
Oops, I damaged a variable name

bhearsum commented:
I haven't looked in detail at these patches yet, but I did push them to our [try server](https://treeherder.mozilla.org/jobs?repo=try&revision=6e57e002c012dd14d73689a6497d3821bb97f480&selectedTaskRun=UE0M0N7CT9SzRAgvcEvEJQ.0) which failed to build it:
```
[task 2025-10-01T00:53:23.068+00:00] cd /builds/worker/fetches/libdmg-hfsplus/hfs && /builds/worker/fetches/clang/bin/clang --sysroot=/builds/worker/fetches/sysroot  -I/builds/worker/fetches/libdmg-hfsplus/includes -O3 -DNDEBUG -fno-strict-aliasing -MD -MT hfs/CMakeFiles/hfs.dir/hfscompress.c.o -MF CMakeFiles/hfs.dir/hfscompress.c.o.d -o CMakeFiles/hfs.dir/hfscompress.c.o -c /builds/worker/fetches/libdmg-hfsplus/hfs/hfscompress.c
[task 2025-10-01T00:53:23.093+00:00] /builds/worker/fetches/libdmg-hfsplus/hfs/hfslib.c:321:13: error: call to undeclared library function 'strlcpy' with type 'unsigned long (char *, const char *, unsigned long)'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
[task 2025-10-01T00:53:23.093+00:00]   321 |         totalLen = strlcpy(fullName, parentName, MAXPATHLEN+1);
[task 2025-10-01T00:53:23.093+00:00]       |                    ^
[task 2025-10-01T00:53:23.093+00:00] /builds/worker/fetches/libdmg-hfsplus/hfs/hfslib.c:321:13: note: include the header <string.h> or explicitly provide a declaration for 'strlcpy'
[task 2025-10-01T00:53:23.109+00:00] 1 error generated.
```

It obviously built fine here, which might mean there's a problem on the Gecko CI side. Notably, those builds are happening on Debian 12 instead of Ubuntu, which might be a factor.

AdamNorberg commented:
Fixing. `strlcpy` is BSD, not POSIX, and some Linux distributions have adopted it while others haven't. Oops! (that's what I get for trusting `strncpy`'s man page when it told me to use `strlcpy` instead.) Will update this when it's fixed and will merge into my branch for PR 21.

AdamNorberg committed:
Switch from strlcpy to strlen + memcpy.

strlcpy, while convenient, is a BSD API not available on the Debian image used for the CI workflow (although the Ubuntu image specified in the Dockerfile is perfectly fine with it). Since strncpy does not provide any way to distinguish between "okay, I copied your entire string" and "btw I couldn't find room for the NUL terminator, hope you didn't need that, lul", I'm using strlen+memcpy explicitly, which is at least more obvious about how it will not help you maintain C string validity in any way.

This implementation relies on the original string to have a \0 in the expected spot, but so does strlen (and all other str-based APIs) so it's not like this makes the situation worse.

AdamNorberg committed:
fix typo

AdamNorberg commented:
Looks okay now?

AdamNorberg commented:
I have had second thoughts about using MAXPATHLEN. That's a system-local limit; I should instead have written a constant in `hfsplus.h`. But it's weird to have that constant and _not_ use it consistently. I'm landing parts of the symlink change (downstream from this one) in Chromium now (finally) and I've changed MAXPATH in that version back to literal 1024s.

If there is any interest from Mozilla folks, I will merge my Chromium changes back into this pull request's branch. The code has benefitted from my teammates' code reviews.

gabrielesvelto commented:
> If there is any interest from Mozilla folks, I will merge my Chromium changes back into this pull request's branch. The code has benefitted from my teammates' code reviews.

Yes please, I'll prioritize reviewing these PRs this week.

gabrielesvelto commented on hfs/btree.c:
```diff
@@ -792,6 +791,7 @@ static int growBTree(BTree* tree) {
     
     while(TRUE) {
       descriptor = readBTNodeDescriptor(mapNode, tree);
+      ASSERT(descriptor, "readBTNodeDescriptor");
```
It would be better to have a descriptive error message, such as "Failed to read BTNodeDescriptor".

AdamNorberg replied:
Done. This assertion was written to be similar to others in the repo; I'll take this as an invitation to clean up error assertion messages to be legible, too! (opportunistically, not going to hunt them all down right now.)

gabrielesvelto commented on hfs/btree.c:
```diff
@@ -826,10 +826,11 @@ static int growBTree(BTree* tree) {
         free(buffer);
-        
+        free(descriptor);
         ASSERT(writeBTHeaderRec(tree), "writeBTHeaderRec");
```
Likewise

AdamNorberg replied:
Done.

gabrielesvelto commented on hfs/xattr.c:
```diff
@@ -251,6 +268,7 @@ size_t getAttribute(Volume* volume, uint32_t fileID, const char* name, uint8_t**
 			return size;
 		default:
 			fprintf(stderr, "unsupported attribute node format\n");
+			free(record);
```
I feel like the code above would be more readable if it were factored out, like:
```
record = (HFSPlusAttrRecord*) search(volume->attrTree, (BTKey*)(&key), &exact, NULL, NULL);
size = MaybeFillAttributeData(exact, record, data);

if (record) {
    free(record);
}

return size;
```
This is just a nit though, the fix itself is sound and I understand you might not want to alter this code too much.

AdamNorberg replied:
Adding the extra helper would be more legible but it's more of a rewrite than I think I want to take right here. I couldn't find anything else that would obviously use MaybeFillAttributeData, so the effort of pulling it out feels a little difficult to justify.

That said, the `switch` statement being used as an if/else is unnecessarily messy. I've refactored it into an early-out check against the record type instead. Is this enough of an improvement here?

gabrielesvelto replied:
Yes, this is fine, thank you!

gabrielesvelto reviewed (changes requested):
Just a handful of nits, the changes look sound.

AdamNorberg committed:
Merge remote-tracking branch 'origin/mozilla' into scattered-substantiative-cleanups

AdamNorberg committed:
Reindent abstractfile.c

AdamNorberg committed:
base64.c needs File, not dmg. switch the include.

This way, "common" can compile as an independent subcomponent (as I've specified Chromium's libdmg-hfsplus build to work); otherwise, the circular dependency requires "dmg" and "common" to always be compiled together.

AdamNorberg committed:
Reindent attribution.c.

AdamNorberg committed:
Reindent checksum.c a little.

AdamNorberg committed:
Reindent dmgfile.c (preferring tabs).

Tabs were predominant in this file,.

AdamNorberg commented:
in progress: pulling in changes I made in Chromium. will address feedback shortly after. thanks!

AdamNorberg committed:
Reindent dmglib.c (with tabs).

Tabs were predominant in this file.

AdamNorberg committed:
Reindent filevault.c. Use braces for one-line ifs.

Let's avoid a `goto fail` of our own.

AdamNorberg committed:
Reindent partition.c, clean up braces.

AdamNorberg committed:
Reindent and rebrace resources.c.

There were substantiative indentation errors in this file. I believe the indentation now matches the behavior.

AdamNorberg committed:
Fix uninitialized read in debugBTree.

AdamNorberg committed:
Fix indentation and braces in catalog.c.

AdamNorberg committed:
Reindent and rebrace volume.c.

AdamNorberg committed:
xattr.c: clean up updateAttributes types, add more braces.

AdamNorberg committed:
Fix indentation and trailing whitespace in dmg.h.

AdamNorberg committed:
Fix hfsplus.h indentation. Delete tryCatalogRecordAsThread family.

Nothing handles those cases anyway, the helper is just clutter. It can easily be added later if needed. (also the underlying file got removed in a catalog.c cleanup that was a paste from a version of the repo where I didn't add the extraneous helpers in the first place, oops)

AdamNorberg committed:
Back out from MAXPATHLEN, it was a bad idea.

The path length limit is an HFS+ thing, not a local system thing, so picking up MAXPATHLEN from POSIX libraries is a mistake; I'd need to add a constant in hfsplib.h instead. But if I'm doing that, I _really_ need to make sure it is consistent through the entire codebase, and I can't justify spending the time on that cleanup right now. So the bounds checks stay, but they will have to rely on length literals.

AdamNorberg committed:
Code review feedback -- assert cleanups, structure cleanups

AdamNorberg commented:
Okay, additional changes from Chromium are now in the patch. Specific things to note:

* hfs/btree.c: debugBTree had an uninitialized read, now fixed.
* catalog.c's change causes hfsplus.h to declare a prototype that is never implemented, but it's also never used. The hfsplus.h change at the end of the Chromium patches chain deletes the prototype. The helpers were new in this patch sequence anyway so it cannot create a conflict with anything else trying to use this library as a library.

Also updated for code review feedback. thank you, @gabrielesvelto!

gabrielesvelto reviewed (approved):
This is fine, thank you! Merging
